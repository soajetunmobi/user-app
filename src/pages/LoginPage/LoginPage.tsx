import React, { FunctionComponent, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { Form, Formik, FormikProps } from 'formik'
import { userService } from 'service'
import { routes } from 'const/routes'
import { loginFormValidation } from 'validation'
import { Button, createStyles, Grid, TextField } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { useUserContext } from '../../providers'

interface ILoginForm {
  username: string
  password: string
}

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      maxWidth: '450px',
      display: 'block',
      margin: '0 auto'
    },
    textField: {
      '& > *': {
        width: '100%'
      }
    },
    submitButton: {
      marginTop: '24px'
    },
    title: { textAlign: 'center' },
    successMessage: { color: 'green' },
    errorMessage: { color: 'red' }
  })
)

const formStatusProps: IFormStatusProps = {
  success: {
    message: 'Logged in successfully. You will be redirected to home page',
    type: 'success'
  },
  fail: {
    message: 'Incorrect username or password.',
    type: 'error'
  },
  error: {
    message: 'Something went wrong. Please try again.',
    type: 'error'
  }
}

const LOGIN_INITIAL_VALUES = {
  username: '',
  password: ''
}

export const LoginPage: FunctionComponent = () => {
  const { setUser } = useUserContext()
  const classes = useStyles()
  const history = useHistory()
  const [displayFormStatus, setDisplayFormStatus] = useState(false)
  const [formStatus, setFormStatus] = useState<IFormStatus>({
    message: '',
    type: ''
  })

  const authenticateUser = async ({ username, password }: ILoginForm) => {
    const loggedInUser = await userService.getByEmail(username)
    if (
      loggedInUser &&
      loggedInUser.email.toLowerCase() === username.toLowerCase() &&
      password.length !== 0
    ) {
      setFormStatus(formStatusProps.success)
      setUser(loggedInUser)
      setTimeout(() => history.push(routes.home), 5000)
    } else {
      setFormStatus(formStatusProps.fail)
    }
    setDisplayFormStatus(true)
  }

  return (
    <div className={classes.root}>
      <Formik
        initialValues={LOGIN_INITIAL_VALUES}
        onSubmit={async (values: ILoginForm, actions) => {
          await authenticateUser(values)
          setTimeout(() => {
            actions.setSubmitting(false)
          })
        }}
        validationSchema={loginFormValidation}
      >
        {(props: FormikProps<ILoginForm>) => {
          const { values, touched, errors, handleBlur, handleChange, isSubmitting } = props
          return (
            <Form data-testid='login-form'>
              <h1 className={classes.title}>Login up</h1>
              <Grid container justify='space-around' direction='row'>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='username'
                    id='username'
                    label='Username'
                    value={values.username}
                    type='email'
                    helperText={
                      errors.username && touched.username ? errors.username : 'Enter username'
                    }
                    error={errors.username && touched.username ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='password'
                    id='password'
                    label='Password'
                    value={values.password}
                    type='password'
                    helperText={
                      errors.password && touched.password ? errors.password : 'Enter password'
                    }
                    error={errors.password && touched.password ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.submitButton}>
                  <Button
                    type='submit'
                    variant='contained'
                    color='secondary'
                    disabled={isSubmitting}
                  >
                    Submit
                  </Button>
                  {displayFormStatus && (
                    <div className='formStatus'>
                      {formStatus.type === 'error' ? (
                        <p className={classes.errorMessage}>{formStatus.message}</p>
                      ) : formStatus.type === 'success' ? (
                        <p className={classes.successMessage}>{formStatus.message}</p>
                      ) : null}
                    </div>
                  )}
                </Grid>
              </Grid>
            </Form>
          )
        }}
      </Formik>
    </div>
  )
}
