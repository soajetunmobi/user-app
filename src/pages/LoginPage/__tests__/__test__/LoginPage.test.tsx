import React from 'react'
import renderer from 'react-test-renderer'
import { LoginPage } from '../../LoginPage'
import { fireEvent, render, waitFor } from '@testing-library/react'

describe('LoginPage', () => {
  it('renders LoginPage correctly', () => {
    const root = renderer.create(<LoginPage />).toJSON()
    expect(root).toMatchSnapshot()
  })

  const usernameMock = 'test@gmail.com'
  const passwordMock = 'test'

  describe('Login form submission', () => {
    it('should show helper text', async () => {
      const { getByText } = render(<LoginPage />)
      expect(getByText('Enter username')).toBeInTheDocument()
      expect(getByText('Enter password')).toBeInTheDocument()
    })

    it('submits correct values', async () => {
      const { container } = render(<LoginPage />)
      const username = container.querySelector('input[name="username"]')
      const password = container.querySelector('input[name="password"]')
      await waitFor(() => {
        fireEvent.change(username, { target: { value: usernameMock } })
        fireEvent.change(password, { target: { value: passwordMock } })
      })
      expect(username.value).toBe(usernameMock)
      expect(password.value).toBe(passwordMock)
    })
  })
})
