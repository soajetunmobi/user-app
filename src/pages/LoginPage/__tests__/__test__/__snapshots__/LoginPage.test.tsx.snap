// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders LoginPage correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <form
    action="#"
    data-testid="login-form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <h1
      className="makeStyles-title-4"
    >
      Login up
    </h1>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-around"
    >
      <div
        className="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
            htmlFor="username"
            id="username-label"
          >
            Username
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-describedby="username-helper-text"
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="username"
              name="username"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="email"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root"
            id="username-helper-text"
          >
            Enter username
          </p>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
            htmlFor="password"
            id="password-label"
          >
            Password
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-describedby="password-helper-text"
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="password"
              name="password"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root"
            id="password-helper-text"
          >
            Enter password
          </p>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-submitButton-3 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Submit
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;
